src/scanner.cc

```diff
@@ -1610,20 +1620,153 @@ namespace {
       }
     }
   };
+  
+  /**
+   * A hierarchy of nested stateful scanners.
+   * Each time a PlusCal block is entered, a nested context is created.
+   * Exiting the PlusCal block exits the context.
+   * Multiply-nested PlusCal blocks are supported.
+   */
+  struct NestedScanner {
+
+    // The enclosing context(s) of the PlusCal block.
+    std::vector<std::vector<char>> enclosing_contexts;
+    
+    // The currently-active context.
+    Scanner current_context;
+
+    /**
+     * Initializes a new instance of the NestedScanner object.
+     */
+    NestedScanner() {
+      this->deserialize(NULL, 0);
+    }
+    
+    unsigned serialize(char* const buffer) {
+      unsigned offset = 0;
+      unsigned copied = 0;
+
+      // First write number of enclosing contexts (guaranteed to be >= 1)
+      nest_address const context_depth = this->enclosing_contexts.size() + 1;
+      copied = sizeof(nest_address);
+      memcpy(&buffer[offset], &context_depth, copied);
+      offset += copied;
+      
+      // Then write size of N-1 enclosing contexts
+      for (int i = 0; i < context_depth - 1; i++) {
+        unsigned const context_size = this->enclosing_contexts[i].size();
+        copied = sizeof(unsigned);
+        memcpy(&buffer[offset], &context_size, copied);
+        offset += copied;
+      }
+      
+      // Reserve space for current context size
+      unsigned const current_context_size_offset = offset;
+      copied = sizeof(unsigned);
+      offset += copied;
+      
+      // Serialize N-1 enclosing contexts
+      for (int i = 0; i < this->enclosing_contexts.size(); i++) {
+        std::vector<char>& context = this->enclosing_contexts[i];
+        copied = context.size();
+        memcpy(&buffer[offset], context.data(), copied);
+        offset += copied;
+      }
+
+      // Serialize current context
+      copied = this->current_context.serialize(&buffer[offset]);
+      offset += copied;
+      
+      // Write current context size to reserved position
+      memcpy(&buffer[current_context_size_offset], &copied, sizeof(unsigned));
```

Comment:
Unnecessary; current context size can be inferred from space remaining in provided byte buffer when deserializing. This means we also don't need the dry run stuff.

Reply:
Actually dry run stuff is used when pushing to the stack.